libsql_client/dbapi2/hrana.py

```diff
@@ -0,0 +1,423 @@
+
+    @run_in_executor
+    def _raw_execute_script(self, sql_script: str) -> FutureNone:
+        assert self._stream is not None
+        return self._stream.sequence(sql_script)
+
+    @run_in_executor
+    def _raw_batch(self, batch: Batch) -> FutureBatchResult:
+        assert self._stream is not None
+        return self._stream.batch(batch)
+
+    def _raw_close(self) -> None:
+        # use of getattr as the object may fail init
+        stream = getattr(self, "_stream", None)
+        if stream is not None:
+            self._destroy_stream(stream)
+            self._stream = None
+
+        conn = getattr(self, "_conn", None)
+        if conn is not None:
+            self._dispose_connection(conn)
+            self._conn = None
+
+        session = getattr(self, "_session", None)
+        if session is not None:
+            self._dispose_session(session)
+            self._session = None
+
+        executor = getattr(self, "_executor", None)
+        if executor is not None:
+            self._dispose_executor(executor)
+            self._executor = None
+
+
+class CursorHrana(Cursor):
+    """Implement :py:class:`sqlite3.Cursor` for remote servers
+    using the `Hrana Protocol
+    <https://github.com/libsql/sqld/blob/main/docs/HRANA_2_SPEC.md>`_.
+    """
+    connection: ConnectionHrana
+
+    def _raw_execute_one(self, stmt: Stmt) -> RawExecuteResult:
+        try:
+            if self.connection._trace_callback is not None:
+                self.connection._trace(stmt["sql"])
+            result = self.connection._raw_execute(stmt)
+            return _conv_stmt_result(result, None)
+        except Exception as error:
+            return _conv_stmt_result(None, error)
+
+    def _raw_execute_multiple(self, stmts: List[Stmt]) -> RawExecuteResult:
+        sql_id = None
+        try:
+            sql = stmts[0]["sql"]
+            sql_id = self.connection._raw_store_sql(sql)
+            assert sql_id is not None  # keep mypy happy
+
+            stored_stmts: List[Stmt] = [
+                _conv_stmt_plain_to_stored(stmt, sql_id) for stmt in stmts
+            ]
+
+            result = self.connection._raw_batch(_conv_batch(stored_stmts))
+            if self.connection._trace_callback is not None:
+                for stmt in stmts:
+                    self.connection._trace(sql)
+            return _conv_batch_result(result)
+        except Exception as error:
+            return _conv_stmt_result(None, error)
+        finally:
+            if sql_id is not None:
+                try:
+                    self.connection._raw_close_sql(sql_id)
+                except Exception:
+                    pass
+
+    def _raw_execute(
+            self,
+            sql: str,
+            parameters: Iterable[SqlParameters],
+            *,
+            want_rows: bool = True,
+    ) -> RawExecuteResult:
+        stmts = _conv_stmts(sql, parameters, want_rows)
+        assert len(stmts) > 0, "Expected at least 1 parameter"
```

Comment:
because it can never happen. I noticed this bug and it happened because I mistyped the default argument, instead of `((),)` it was `()`.

decided to leave this assertion, note that running python with `-OO` will remove assertions.

Reply:
So calling `cursor.executemany("SELECT ?", [])` is forbidden?

Reply:
while `executemany()` doesn't allow SELECT, it does allow such empty list for insert... fixed.

however I also noticed that the behavior cannot be 100% matching since sqlite3 will prepare the statement even if no parameters are given... then you can get exceptions if they are invalid.

Reply:
You can ensure that the statement is prepared (and exceptions are raised) by sending a `describe` Hrana request if the list is empty, but that's probably not necessary.